File: Python/295. FindMedianFromDataStream.py
"""
Problem:
--------

Design a data structure that supports the following two operations:

- `void addNum(int num)`: Add a integer number from the data stream to the data structure.
- `double findMedian()`: Return the median of all elements so far.
"""


class MedianFinder:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.list = []

    def addNum(self, num: int) -> None:
        # Traverse through the list and check if `num` > ith element
        # If yes, insert `num` in that index
        # This keeps the list sorted at all times
        for i in range(len(self.list)):
            if num > self.list[i]:
                self.list.insert(i, num)
                return

        # If `num` is the largest element or is the first one to be added
        self.list.append(num)

    def findMedian(self) -> float:
        # Find index of the middle element (floor division by 2)
        mid_index = len(self.list) // 2

        if len(self.list) % 2 == 0:
            # If number of elements = EVEN
            # Return average of the middle 2 elements
            return (self.list[mid_index - 1] + self.list[mid_index]) / 2
        else:
            # If number of elements = ODD
            # Return the middle element
            return self.list[mid_index]


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: C++/295. FindMedianfromDataStream.cpp
/*
    We maintain 2 heaps with equal or almost equal sizes. meaning the size between one heap
    from another will have a difference of at most 1 item. We call them left heap and right heap
    so that the middle will always be the median. Then we add the items accordingly depending
    on the top item of the left numheap. Keep adding them and the top of the left heap will contain
    the median in the case of an odd numbered size, and both left and right heap top will be the median
    when there is an even numbered size.
*/

class MedianFinder {
    priority_queue<int> left;                              // maxheap
    priority_queue<int, vector<int>, greater<int>> right;  // minheap
   public:
    /** initialize your data structure here. */
    MedianFinder() {
        // do nothing.
    }

    void addNum(int num) {
        if (left.size() == 0 || num < left.top()) {
            left.push(num);
        } else {
            right.push(num);
        }

        if (right.size() > left.size()) {
            left.push(right.top());
            right.pop();
        } else if (left.size() - 1 > right.size()) {
            right.push(left.top());
            left.pop();
        }
    }

    double findMedian() {
        int combinedSize = left.size() + right.size();
        if (combinedSize == 0) {
            return 0;
        }

        if (combinedSize % 2 == 0) {
            return ((double)left.top() + (double)right.top()) / (double)2;
        } else {
            return left.top();
        }
    }
};